ssh-audit.py

```diff
@@ -3713,14 +3720,55 @@ def audit(aconf: AuditConf, sshv: Optional[int] = None, print_target: bool = Fal
 
     return program_retval
 
+def algorithm_lookup(alg_names):       
+    if not alg_names:    
+        out.fail("[exception] Missing argument: alg_names")
+        sys.exit(1)
+        
+    alg_types = {'kex': 'key exchange algorithms', 
+                 'key': 'host-key algorithms',
+                 'mac': 'message authentication code algorithms',
+                 'enc': 'encryption algorithms (ciphers)'
+    }
+    
+    algorithms = alg_names.split(",")
+    adb = SSH2.KexDB.ALGORITHMS
+    algorithms_dict = {}
+    
+    for alg_type in alg_types:
+        for alg_name in algorithms:
+            try:
+                SSH2.KexDB.ALGORITHMS[alg_type][alg_name]
+                algorithms_dict.update( {alg_name : alg_type} ) 
+            except Exception:
+                pass
+
+    unknown_algorithms = []
+    padding = len(max(algorithms, key=len))
+        
+    for alg_type in alg_types:
+        algorithms_dict_filtered = {k:v for (k,v) in algorithms_dict.items() if alg_type in v}
+        
+        if len(algorithms_dict_filtered.items()) > 0:
+            title = alg_types.get(alg_type)
+            output_algorithms(title, adb, alg_type, algorithms_dict_filtered.keys(), unknown_algorithms, False, PROGRAM_RETVAL_GOOD, padding)
+          
+    algorithms_not_found = [i for i in algorithms if i not in algorithms_dict]
+    if len(algorithms_not_found) > 0:
+        out.head('# unknown algorithms')
+        for algorithm_not_found in algorithms_not_found:
+            out.fail(algorithm_not_found) 
```

Comment:
I suppose `retval = PROGRAM_RETVAL_FAILURE` would be appropriate when the algorithm isn't found.

Reply:
@jtesta Would you like `PROGRAM_RETVAL_FAILURE` to be returned if any one of the specified algorithms could not be found? Or only if **all** of the specified algorithms could not be found?

Reply:
How about if all the specified algorithms could not be found then return `PROGRAM_RETVAL_FAILURE`, if some (but not all) of the specified algorithms could not be found then return `PROGRAM_RETVAL_WARNING`, else return `PROGRAM_RETVAL_GOOD` ?

Reply:
The way it works now, if at least one algorithm fails, the program returns failure overall (it doesn't matter if all other algs are good).

So perhaps we do the same thing here: if at least one is unknown (= failure), the result should be failure overall.